nova/virt/vmwareapi/volumeops.py

```diff
@@ -590,3 +591,74 @@ def attach_root_volume(self, connection_info, instance,
         # we don't do that. This comment should help us detect upstream changes
         # to the function as merging should fail.
         self.attach_volume(connection_info, instance, adapter_type)
+
+    def fixup_shadow_vms(self, instance, shadow_vms):
+        # This should sensibly moved out to cinder
```

Comment:
It doesn't need to. It can delete the shadow-vm, when the attachment for the old vm/shadow-vm gets deleted.

Reply:
For that to work, it must know that the volume migrated. We cannot just let Cinder delete shadow-VMs :D

Reply:
That's kind of the point of the terminate connection call in this context. (Having more than one connection / attachments to a VM)

* After a successful migration, a terminate connection call will be issued with the connection-info related to the "old" shadow-vm. Since successful implies a migration of the VM and its volumes, the data has been safely migrated, and the old shadow-vm is obsolete. Instead of Nova deleting the shadow-vms there Cinder can do the necessary steps.

* After a roll-back, a terminate connection associated with the new shadow-vm will be called, so the migration was not successful, and the new shadow-vm does not point to the new data, so we can get rid of it. And instead of Nova doing it, it would be better encapsulated in Cinder.

Reply:
hm ... ok, so Cinder sees that Nova provides "not matching" connection_info and thus thinks "well, the migration must have succeeded, let's delete the shadow-vm"? Might work.

Reply:
Not quite. We have a 1:1 relationship between attachments/connections and shadow-vm. And having more than one  (two) is only valid during a live-migration.
Cinder gets a termination call for a connection, and since we have too many shadow-vms, that termination call tells us exactly which shadow-vm of the two has to go. It is the one whose connection has been terminated.

That works irregardless of migration having succeeded for failed.